GeneratorInterface/HydjetInterface/interface/HydjetHadronizer.h

```diff
@@ -32,7 +36,7 @@ namespace HepMC {
 namespace gen {
   class Pythia6Service;
 
-  class HydjetHadronizer : public BaseHadronizer {
+  class HydjetHadronizer : public BaseHadronizer, public edm::EDConsumerBase {
```

Comment:
This does not work as intended. The framework will not magically get the `consumes()` information by arbitrary components declaring to inherit from `edm::EDConsumerBase`. The `edm::ConsumesCollector` should be used instead to call `consumes()` (it needs to be percolated to the constructor of `HydjectHadronizer` from the constructors of all the EDModules that use it).

Reply:
@makortel I assumed to use **e.setConsumer(this)** after [getEDMEvent()](https://github.com/cms-sw/cmssw/blob/a6eede42db53e5bef3393ba46886f91e38d25812/GeneratorInterface/HydjetInterface/src/HydjetHadronizer.cc#L195) or setEDMEvent() (which we have to add somewhere). I tests it looks worked, but I can be wrong...

Reply:
That is not sufficient (anyway `Event::setConsumer()` mostly framework internal function). Framework needs to know the consumes information (=data dependencies) at the module construction time.

Reply:
@makortel [Isn't it](https://github.com/cms-sw/cmssw/blob/a6eede42db53e5bef3393ba46886f91e38d25812/GeneratorInterface/HydjetInterface/src/HydjetHadronizer.cc#L60)? 
Sorry if I'm being stupid, but could You please to explain Your suggestion to replace  `edm::EDConsumerBase` by `edm::ConsumesCollector`? I mean how to implement it?

Reply:
> [Isn't it](https://github.com/cms-sw/cmssw/blob/a6eede42db53e5bef3393ba46886f91e38d25812/GeneratorInterface/HydjetInterface/src/HydjetHadronizer.cc#L60)?

No. The call to the `EDConsumerBase` constructor is far from sufficient. In short the framework asks from the `EDConsumerBase` objects **it knows** (=EDModules constructed by the framework) which products are consumed. The framework **is not aware** of arbitrary classes that decide to inherit from `EDConsumerBase`.

> Sorry if I'm being stupid, but could You please to explain Your suggestion to replace `edm::EDConsumerBase` by `edm::ConsumesCollector`? I mean how to implement it?

See https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideEDMGetDataFromEvent#Consumes_and_Helpers

Essentially it should be something along the EDModule using `HydjectHadronizer`, e.g. `edm::GeneratorFilter<HAD, DEC>` doing
```cpp
  template <class HAD, class DEC>
  GeneratorFilter<HAD, DEC>::GeneratorFilter(ParameterSet const& ps)
      : EDFilter(), hadronizer_(ps, consumesCollector()), decayer_(nullptr), nEventsInLumiBlock_(0) {
```
and then `HydjectHadronizer` doing
```cpp
HydjetHadronizer::HydjetHadronizer(const ParameterSet& pset, ConsumesCollector iC)
... {
  ...
  if (embedding_) {
    cflag_ = 0;
    src_ = iC.consumes<HepMCProduct>(
        pset.getUntrackedParameter<edm::InputTag>("backgroundLabel", edm::InputTag("generator", "unsmeared")));
  }
  ...
}

```
And yes, it does imply that *all* hadronizers that are used in conjunction with `edm::GeneratorFilter<...>` need to be modified to take `edm::ConsumesCollector` argument (that they can ignore).

Reply:
@makortel Thanks! I works, but with an exception:

```
Exception Message:
Unrunnable schedule
Module run order problem found: 
generatorSmeared after VtxSmeared [path generation_step], VtxSmeared after mix [path generation_step], mix after randomEngineStateProducer [path generation_step], randomEngineStateProducer after generator [path generation_step], generator consumes generatorSmeared
 Running in the threaded framework would lead to indeterminate results.
 Please change order of modules in mentioned Path(s) to avoid inconsistent module ordering.

```
I think we have firstly solve the [issue](https://github.com/cms-sw/cmssw/issues/28636).

Reply:
> It works,

Good.

> I think we have firstly to solve the [issue](https://github.com/cms-sw/cmssw/issues/28636).

Correct.

Reply:
> > [Isn't it](https://github.com/cms-sw/cmssw/blob/a6eede42db53e5bef3393ba46886f91e38d25812/GeneratorInterface/HydjetInterface/src/HydjetHadronizer.cc#L60)?
> 
> No. The call to the `EDConsumerBase` constructor is far from sufficient. In short the framework asks from the `EDConsumerBase` objects **it knows** (=EDModules constructed by the framework) which products are consumed. The framework **is not aware** of arbitrary classes that decide to inherit from `EDConsumerBase`.
> 
> > Sorry if I'm being stupid, but could You please to explain Your suggestion to replace `edm::EDConsumerBase` by `edm::ConsumesCollector`? I mean how to implement it?
> 
> See https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideEDMGetDataFromEvent#Consumes_and_Helpers
> 
> Essentially it should be something along the EDModule using `HydjectHadronizer`, e.g. `edm::GeneratorFilter<HAD, DEC>` doing
> 
> ```c++
>   template <class HAD, class DEC>
>   GeneratorFilter<HAD, DEC>::GeneratorFilter(ParameterSet const& ps)
>       : EDFilter(), hadronizer_(ps, consumesCollector()), decayer_(nullptr), nEventsInLumiBlock_(0) {
> ```
> 

I changed all generators for this, but have a trouble with Pythia8:

'''
>> Compiling edm plugin /afs/cern.ch/work/w/wouf/HYDJET/1.9.1/EmbeddingProj/CMSSW_11_1_0_pre5/src/GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc 
In file included from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/external/dire/2.004-bcolbf/include/Dire/SplitInfo.h:17,
                 from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/external/dire/2.004-bcolbf/include/Dire/Splittings.h:15,
                 from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/external/dire/2.004-bcolbf/include/Dire/SplittingLibrary.h:15,
                 from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/external/dire/2.004-bcolbf/include/Dire/Dire.h:8,
                 from /afs/cern.ch/work/w/wouf/HYDJET/1.9.1/EmbeddingProj/CMSSW_11_1_0_pre5/src/GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc:15:
/cvmfs/cms.cern.ch/slc7_amd64_gcc820/external/dire/2.004-bcolbf/include/Dire/DireBasics.h:108:7: warning: 'class Pythia8::DireFunction' has virtual functions and accessible non-virtual destructor [-Wnon-virtual-dtor]
 class DireFunction {
       ^~~~~~~~~~~~
In file included from /afs/cern.ch/work/w/wouf/HYDJET/1.9.1/EmbeddingProj/CMSSW_11_1_0_pre5/src/GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc:61:
/afs/cern.ch/work/w/wouf/HYDJET/1.9.1/EmbeddingProj/CMSSW_11_1_0_pre5/src/GeneratorInterface/Core/interface/ConcurrentHadronizerFilter.h: In instantiation of 'edm::gen::StreamCache<HAD, DEC>::StreamCache(const edm::ParameterSet&) [with HAD = Pythia8Hadronizer; DEC = gen::ConcurrentExternalDecayDriver]':
/cvmfs/cms.cern.ch/slc7_amd64_gcc820/external/gcc/8.2.0-pafccj/include/c++/8.3.1/bits/unique_ptr.h:831:30:   required from 'typename std::_MakeUniq<_Tp>::__single_object std::make_unique(_Args&& ...) [with _Tp = edm::gen::StreamCache<Pythia8Hadronizer, gen::ConcurrentExternalDecayDriver>; _Args = {const edm::ParameterSet&}; typename std::_MakeUniq<_Tp>::__single_object = std::unique_ptr<edm::gen::StreamCache<Pythia8Hadronizer, gen::ConcurrentExternalDecayDriver>, std::default_delete<edm::gen::StreamCache<Pythia8Hadronizer, gen::ConcurrentExternalDecayDriver> > >]'
/afs/cern.ch/work/w/wouf/HYDJET/1.9.1/EmbeddingProj/CMSSW_11_1_0_pre5/src/GeneratorInterface/Core/interface/ConcurrentHadronizerFilter.h:176:62:   required from 'std::unique_ptr<edm::gen::StreamCache<HAD, DEC> > edm::ConcurrentHadronizerFilter<HAD, DEC>::beginStream(edm::StreamID) const [with HAD = Pythia8Hadronizer; DEC = gen::ConcurrentExternalDecayDriver]'
/afs/cern.ch/work/w/wouf/HYDJET/1.9.1/EmbeddingProj/CMSSW_11_1_0_pre5/src/GeneratorInterface/Core/interface/ConcurrentHadronizerFilter.h:175:47:   required from here
/afs/cern.ch/work/w/wouf/HYDJET/1.9.1/EmbeddingProj/CMSSW_11_1_0_pre5/src/GeneratorInterface/Core/interface/ConcurrentHadronizerFilter.h:70:65: error: no matching function for call to 'Pythia8Hadronizer::Pythia8Hadronizer(<brace-enclosed initializer list>)'
       StreamCache(ParameterSet const& iPSet) : hadronizer_{iPSet} {}
                                                                 ^
/afs/cern.ch/work/w/wouf/HYDJET/1.9.1/EmbeddingProj/CMSSW_11_1_0_pre5/src/GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc:170:1: note: candidate: 'Pythia8Hadronizer::Pythia8Hadronizer(const edm::ParameterSet&, edm::ConsumesCollector&&)'
 Pythia8Hadronizer::Pythia8Hadronizer(const edm::ParameterSet &params, edm::ConsumesCollector&& iC)
 ^~~~~~~~~~~~~~~~~
/afs/cern.ch/work/w/wouf/HYDJET/1.9.1/EmbeddingProj/CMSSW_11_1_0_pre5/src/GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc:170:1: note:   candidate expects 2 arguments, 1 provided
''' 

@makortel @mkirsano Could You please help to fix it?